File: packages/host-ios/Sources/ComicalHostIOS/ComicalServer.swift
/**
 * ComicalServer — optional on-device HTTP server for iOS.
 *
 * For apps that want to expose the bridge runtime as a local API (so a WKWebView or other
 * client on the same device can call it), this wraps ComicalBridgeContext in a minimal
 * HTTP server using Network.framework. Port is dynamically assigned; the app can pass it to
 * a WKWebView via a custom URL scheme or injected JS.
 *
 * This is equivalent to comical-server (the Bun host-server), but running entirely on-device
 * with no external infrastructure.
 *
 * NOTE: for most iOS apps the better model is to use ComicalBridgeContext directly from Swift
 * code via the `call(_:args:)` API. ComicalServer is for hybrid apps with a web UI.
 */
import Foundation
import Network
import os

public final class ComicalServer {
    private let context: ComicalBridgeContext
    private var listener: NWListener?
    private let log = Logger(subsystem: "dev.comical", category: "server")
    public private(set) var port: UInt16 = 0

    public init(context: ComicalBridgeContext) {
        self.context = context
    }

    public func start() throws {
        let params = NWParameters.tcp
        listener = try NWListener(using: params, on: 0)
        listener?.newConnectionHandler = { [weak self] conn in
            self?.handleConnection(conn)
        }
        listener?.stateUpdateHandler = { [weak self] state in
            if case .ready = state {
                self?.port = self?.listener?.port?.rawValue ?? 0
                self?.log.info("comical on-device server on :\(self?.port ?? 0)")
            }
        }
        listener?.start(queue: .global(qos: .utility))
    }

    public func stop() {
        listener?.cancel()
        listener = nil
    }

    // MARK: - Request handling

    private func handleConnection(_ conn: NWConnection) {
        conn.start(queue: .global(qos: .utility))
        receiveRequest(on: conn)
    }

    private func receiveRequest(on conn: NWConnection) {
        conn.receive(minimumIncompleteLength: 1, maximumLength: 65536) { [weak self] data, _, isComplete, error in
            guard let self, let data, !data.isEmpty else { conn.cancel(); return }
            Task {
                let response = await self.handleHTTP(data: data)
                conn.send(content: response, completion: .contentProcessed { _ in conn.cancel() })
            }
        }
    }

    private func handleHTTP(data: Data) async -> Data {
        guard let req = String(data: data, encoding: .utf8) else {
            return httpResponse(status: 400, body: #"{"error":"invalid request"}"#)
        }

        let lines = req.components(separatedBy: "\r\n")
        guard let firstLine = lines.first else {
            return httpResponse(status: 400, body: #"{"error":"empty request"}"#)
        }
        let parts = firstLine.components(separatedBy: " ")
        guard parts.count >= 2 else {
            return httpResponse(status: 400, body: #"{"error":"malformed request line"}"#)
        }
        let path = parts[1]

        do {
            let result = try await route(path: path)
            return httpResponse(status: 200, body: result)
        } catch {
            return httpResponse(status: 500, body: #"{"error":"\#(error.localizedDescription)"}"#)
        }
    }

    private func route(path: String) async throws -> String {
        let url = URL(string: "http://localhost\(path)")!
        let query = URLComponents(url: url, resolvingAgainstBaseURL: false)?.queryItems ?? []
        let p = url.path

        if p == "/health" { return #"{"ok":true}"# }

        if p == "/info" {
            guard let info = context.bridgeInfo,
                  let data = try? JSONEncoder().encode(info),
                  let json = String(data: data, encoding: .utf8)
            else { throw ComicalError(message: "bridge info unavailable") }
            return json
        }

        if p == "/search" {
            let q = query.first(where: { $0.name == "q" })?.value ?? ""
            let page = Int(query.first(where: { $0.name == "page" })?.value ?? "1") ?? 1
            return try await callBridge("getSearchResults", args: [q, page])
        }

        if p == "/home" { return try await callBridge("getHomeSections") }
        if p == "/popular" {
            let page = Int(query.first(where: { $0.name == "page" })?.value ?? "1") ?? 1
            return try await callBridge("getPopular", args: [page])
        }

        // /series/:id
        if let m = p.range(of: #"^/series/([^/]+)/chapters/([^/]+)/pages$"#, options: .regularExpression) {
            let comps = String(p[m]).components(separatedBy: "/").filter { !$0.isEmpty }
            return try await callBridge("getChapterPages", args: [comps[1], comps[3]])
        }
        if let m = p.range(of: #"^/series/([^/]+)/chapters$"#, options: .regularExpression) {
            let id = String(p[m]).components(separatedBy: "/")[2]
            return try await callBridge("getChapters", args: [id])
        }
        if let m = p.range(of: #"^/series/([^/]+)$"#, options: .regularExpression) {
            let id = String(p[m]).components(separatedBy: "/")[2]
            return try await callBridge("getSeriesDetails", args: [id])
        }

        throw ComicalError(message: "unknown route: \(p)")
    }

    private func callBridge(_ method: String, args: [Any] = []) async throws -> String {
        let result = try await context.call(method, args: args)
        guard let data = try? JSONSerialization.data(withJSONObject: result),
              let json = String(data: data, encoding: .utf8)
        else { throw ComicalError(message: "failed to serialise result of \(method)") }
        return json
    }

    private func httpResponse(status: Int, body: String) -> Data {
        let bodyData = body.data(using: .utf8) ?? Data()
        let header = "HTTP/1.1 \(status) OK\r\nContent-Type: application/json\r\nContent-Length: \(bodyData.count)\r\nAccess-Control-Allow-Origin: *\r\n\r\n"
        return (header.data(using: .utf8) ?? Data()) + bodyData
    }
}

File: packages/host-android/src/main/kotlin/dev/comical/host/ComicalServer.kt
/**
 * ComicalServer — optional on-device HTTP server for Android.
 *
 * Equivalent to ComicalServer.swift — exposes the bridge runtime as a local HTTP API,
 * useful for hybrid apps where a WebView needs to call bridge methods via fetch().
 * Uses NanoHTTPD as a minimal embedded HTTP server (no Play Services dependency).
 *
 * For most apps, use ComicalBridgeContext.call() directly from Kotlin.
 */
package dev.comical.host

import kotlinx.coroutines.runBlocking
import org.json.JSONArray
import org.json.JSONObject
import java.io.InputStream
import java.net.InetSocketAddress
import java.net.ServerSocket
import java.net.Socket
import kotlin.concurrent.thread

class ComicalServer(private val context: ComicalBridgeContext) {
    private var serverSocket: ServerSocket? = null
    var port: Int = 0
        private set

    fun start() {
        val ss = ServerSocket()
        ss.bind(InetSocketAddress("127.0.0.1", 0))
        port = ss.localPort
        serverSocket = ss
        thread(isDaemon = true) {
            while (!ss.isClosed) {
                try {
                    val conn = ss.accept()
                    thread(isDaemon = true) { handle(conn) }
                } catch (_: Exception) {}
            }
        }
    }

    fun stop() { serverSocket?.close() }

    private fun handle(conn: Socket) {
        try {
            val input = conn.getInputStream().bufferedReader()
            val firstLine = input.readLine() ?: return
            // Drain headers
            var line: String
            do { line = input.readLine() ?: break } while (line.isNotEmpty())

            val parts = firstLine.split(" ")
            val path = if (parts.size >= 2) parts[1] else "/"

            val (status, body) = runBlocking {
                try { 200 to route(path) }
                catch (e: Exception) { 500 to """{"error":"${e.message}"}""" }
            }

            val response = buildString {
                append("HTTP/1.1 $status OK\r\n")
                append("Content-Type: application/json\r\n")
                append("Access-Control-Allow-Origin: *\r\n")
                append("Content-Length: ${body.toByteArray().size}\r\n")
                append("\r\n")
                append(body)
            }
            conn.getOutputStream().write(response.toByteArray())
        } finally {
            conn.close()
        }
    }

    private suspend fun route(path: String): String {
        val idx = path.indexOf('?')
        val p = if (idx >= 0) path.substring(0, idx) else path
        val query = parseQuery(if (idx >= 0) path.substring(idx + 1) else "")

        return when {
            p == "/health" -> """{"ok":true}"""
            p == "/info" -> toJson(context.call("info"))
            p == "/search" -> toJson(context.call("getSearchResults", listOf(query["q"] ?: "", (query["page"] ?: "1").toInt())))
            p == "/lists" -> toJson(context.call("getLists"))
            Regex("^/lists/([^/]+)$").matches(p) ->
                toJson(context.call("getListItems", listOf(p.split("/")[2], (query["page"] ?: "1").toInt())))
            Regex("^/series/([^/]+)/chapters/([^/]+)/pages$").matches(p) -> {
                val parts = p.split("/")
                toJson(context.call("getChapterPages", listOf(parts[2], parts[4])))
            }
            Regex("^/series/([^/]+)/chapters$").matches(p) -> toJson(context.call("getChapters", listOf(p.split("/")[2])))
            Regex("^/series/([^/]+)$").matches(p) -> toJson(context.call("getSeriesDetails", listOf(p.split("/")[2])))
            else -> throw ComicalError("unknown route: $p")
        }
    }

    private fun toJson(value: Any): String = when (value) {
        is JSONObject -> value.toString()
        is JSONArray  -> value.toString()
        is String     -> value
        else          -> value.toString()
    }

    private fun parseQuery(query: String): Map<String, String> =
        query.split("&").mapNotNull {
            val eq = it.indexOf('=')
            if (eq < 0) null else it.substring(0, eq) to it.substring(eq + 1)
        }.toMap()
}
